File: backend/models/data_loader.js
const fs = require("fs");
const path = require("path");
const csv = require("csv-parser");

function loadDataset(fileName) {
  const questions = [];
  const filePath = path.join(__dirname, fileName);

  return new Promise((resolve, reject) => {
    fs.createReadStream(filePath)
      .pipe(csv())
      .on("data", (row) => {
        // Normalize keys by trimming and accessing properly
        questions.push({
          topic: row["Topic"],
          question: row["Question"],
          optionA: row["Option A"],
          optionB: row["Option B"],
          optionC: row["Option C"],
          optionD: row["Option D"],
          correctAnswer: row["Correct Answer"],
          difficulty: row["Difficulty"],
          explanation: row["Explanation"],
        });
      })
      .on("end", () => {
        resolve(questions);
      })
      .on("error", (error) => {
        reject(error);
      });
  });
}

// Load and print the dataset
const fileName = "../models/questions_dataset.csv";
loadDataset(fileName)
  .then((dataset) => {
    console.log("Dataset loaded successfully!");
    dataset.slice(0, 5).forEach((question, index) => {
      console.log(
        `${index + 1}. ${question.question} (Topic: ${
          question.topic
        }, Difficulty: ${question.difficulty})`
      );
    });
  })
  .catch((error) => {
    console.error("Error loading dataset:", error.message);
  });

File: backend/models/data_loader.py
import csv

def load_dataset(file_name):
    try:
        with open(file_name, mode='r', encoding='utf-8') as file:
            reader = csv.DictReader(file)
            questions = list(reader)
            return questions
    except FileNotFoundError:
        print(f"Error: The file '{file_name}' was not found.")
        return None

# Specify the file name
file_name = "questions_dataset.csv"

# Load and print the dataset
dataset = load_dataset(file_name)
if dataset:
    print("Dataset loaded successfully!")
    for i, question in enumerate(dataset[:5], start=1):  # Display the first 5 questions
        print(f"{i}. {question['question']} (Topic: {question['topic']}, Difficulty: {question['difficulty']})")
